signing-with-metamask/index.md

```diff
@@ -0,0 +1,89 @@
+2019: MetaMask introduced the ability to sign and send transactions directly from the browser extension, eliminating the need for users to copy and paste the transaction data into a separate signing tool.
+2020: MetaMask introduced the ability to sign transactions using hardware wallets, such as Trezor and Ledger, providing users with an additional layer of security.
+2021: MetaMask introduced the ability to sign messages using a new, more secure method called the "Ethereum Message Signing Protocol (EMSP)".
+
+MetaMask has continued to evolve and improve throughout its journey, providing users with more secure and user-friendly ways to sign data and transactions on the Ethereum blockchain. With the latest updates, MetaMask is now one of the most popular tools for signing data and transactions on the Ethereum network, used by developers and users alike.
+
+Using ethers.js for signing data and transactions on the Ethereum blockchain with MetaMask is convenient, providing an alternative approach. It offers a developer-friendly API and better support for TypeScript, which makes it easier for developers to build dApps and interact with the Ethereum blockchain.
+
+While MetaMask can provide a user-friendly interface for signing data and transactions, ethers.js, an alternative to web3.js (another JavaScript library for interacting with the Ethereum blockchain), has been focused on providing a developer-friendly library for building dApps and interacting with the Ethereum blockchain. Developers have widely adopted ethers.js as a more intuitive and easy-to-use library for signing data and transactions.
+
+MetaMask and ethers.js are not mutually exclusive and can be used together; developers can use MetaMask to interact with the Ethereum blockchain and sign data and transactions, and use ethers.js to build their dApp's front-end. ethers.js may also be used to sign data and transactions on other Ethereum clients such as Geth or Parity.
+
+## Methods of Signing Data With MetaMask
+
+We'll explore various ways of signing data using the provider injected by MetaMask as well as implementations in web3.js and ethers.js:
+
+eth_sign: This method is used to sign data with the Ethereum address's private key and returns the signature in the format of `"0x<r><s><v>"`.
+
+The `eth_sign` method is used to sign data and prove ownership of an Ethereum address. A limitation of this method is that it does not include any prefix before signing the data, which could lead to the signature not being recognized as valid by some smart contracts. It also poses a security risk, so it should be avoided.
+
+personal_sign: This method is also used to sign data and prove ownership of an Ethereum address, similar to `eth_sign,` but adds an "Ethereum Signed Message" prefix to the data before signing it. It is part of the JSON-RPC personal module, which is for personal account management. It signs data with a user's private key and returns the signature. 
```

Comment:
This might be confusing for developers that are also working with other wallets, because `personal_sign` isn't just similar to `eth_sign` - it _is_ `eth_sign`, or at least the updated standardized form of it.

Our `eth_sign` follows an older specification for that signing method. It was updated to use the "Ethereum Signed Message" prefix, but we left our implementation without that prefix to avoid breaking dapps that relied on it. 

Instead we adopted `personal_sign` but implemented it like the modern `eth_sign`. This is confusing as well, because `personal_sign` has a spec that we're not following as well! It's specified to be identical to the modern `eth_sign` except that it also accepts a password for the account, but we don't accept any password.

So someone coming from a different wallet might:

* Expect that `eth_sign` includes a prefix, but ours does not
* Expect `eth_sign` to be the standard way of signing a plain string, but for us it's `personal_sign`,
* Expect `personal_sign` to accept a password, but it does not

This will probably be tricky to explain. Not exactly sure how to approach it, or how much detail to include. For a new dev, fewer details would be better surely.